core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+    class TarFileReadInput(base.BaseInteraction):
+        """Interaction for uploading a file."""
+
+        name = 'Tar file upload'
+        description = 'A widget for uploading tar files.'
+        display_mode = base.DISPLAY_MODE_INLINE
+        _dependency_ids = []
+        answer_type = 'TarFileString'
+        instructions = None
+        narrow_instructions = None
+        needs_summary = False
+        _customization_arg_specs = []
+
+    @classmethod
+    def _ensure_removed_interactions_are_in_registry(cls):
+        """Checks if the FileReadInput or TarFileReadInput deleted interactions
+        are present in the interaction_registry. If they are not, placeholder
+        versions of them will be injected for the purposes of migrating answers
+        from explorations that used to contain them.
+        """
+        interactions = interaction_registry.Registry.get_all_interactions()
+        interaction_names = [
+            type(interaction).__name__ for interaction in interactions]
+        if 'FileReadInput' not in interaction_names:
+            interaction_registry.Registry._interactions['FileReadInput'] = (
+                cls.FileReadInput())
+        if 'TarFileReadInput' not in interaction_names:
+            interaction_registry.Registry._interactions['TarFileReadInput'] = (
+                cls.TarFileReadInput())
+
+        interaction_registry.Registry.get_interaction_by_id('FileReadInput')
+        interaction_registry.Registry.get_interaction_by_id('TarFileReadInput')
+
+    @classmethod
+    def _get_exploration_models_by_versions(cls, exp_id, versions):
+        """Similar to VersionedModel.get_version(), except this allows retrieval
+        of exploration models marked as deleted.
+        """
+        try:
+            # pylint: disable=protected-access
+            snapshot_ids = [
+                exp_models.ExplorationModel._get_snapshot_id(exp_id, version)
+                for version in versions]
+            exp_models_by_versions = [
+                exp_models.ExplorationModel(id=exp_id) for _ in versions]
+            return (exp_models.ExplorationModel._reconstitute_multi_from_models(
+                zip(snapshot_ids, exp_models_by_versions)), None)
+        except Exception as error:
+            return (None, error)
+
+    @classmethod
+    def _force_rule_spec_subjects_to_answer(cls, states_dict):
+        for state_dict in states_dict.values():
+            # Only MultipleChoiceInput interactions with non-answer subjects are
+            # supported for migration.
+            if state_dict['interaction']['id'] != 'MultipleChoiceInput':
+                continue
+            for handler_dict in state_dict['interaction']['handlers']:
+                for rule_spec_dict in handler_dict['rule_specs']:
+                    definition_dict = rule_spec_dict['definition']
+                    param_changes = (
+                        rule_spec_dict['param_changes']
```

Comment:
It probably is unsued before this migration, but I'm keeping these here regardless.